<a href="https://colab.research.google.com/github/brianshin12/zeta/blob/master/Chapter_5%26Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 5장 연습문제

 Q1. 다음은 Calculator 클래스이다.  
 
      class Calculator:  
        def __init__(self):  
          self.value=0  
        def add(self, val):
          self.value += val

위 클래스를 상속하는 UpgradeCalculator를 만들고 값을 뺄 수 있는 minus 메서드를 추가해 보자.  
즉 다음과 같이 동작하는 클래스를 만들어야 한다.  

      cal = UpgradeCalculator()  
      cal.add(10)  
      cal.minus(7)  

      print(cal.value)

In [0]:
 class Calculator:  
    def __init__(self):  
      self.value=0  
    def add(self, val):
      self.value += val

In [2]:
class UpgradeCalculator(Calculator):
  def minus (self, val):
    self.value -= val

cal=UpgradeCalculator()
cal.add(10)
cal.minus(7)

print(cal.value)

3


Q2. 객체변수 value가 100 이상의 값은 가질 수 없도록 제한하는 MaxLimitCalculator 클래스를 만들어 보자. 즉 다음과 같이 동작해야 한다.  

    cal=MaxLimitCalculator()  
    cal.add(50) # 50 더하기  
    cal.add(60) # 60 더하기  
  
    print(cal.value) # 100 출력  

단 반드시 다음과 같은 Calculator 클래스를 상속해서 만들어야 한다.  

    class Calculator:  
      def __init__(self):
        self. value = 0

      def add(self, val):
        self.value += val

In [0]:
class MaxLimitCalculator(Calculator):
  def add(self, val):
    self.value += val
    if self.value >= 100:
      self.value = 100

In [7]:
cal=MaxLimitCalculator()
cal.add(50)
cal.add(60)

print(cal.value)

100


Q3. 다음 결과를 예측해보자.  
1. all([1,2,abs(-3)-3])  
2. chr(ord('a')) == 'a'  

* False  
* True

In [8]:
all([1,2,abs(-3)-3])

False

In [10]:
chr(ord('a'))=='a'

True

Q4. filter와 lambda를 사용하여 리스트 [1, -2, 3, -5, 8, -3]에서 음수를 모두 제거해 보자.

In [11]:
list(filter(lambda x:x>0, [1, -2, 3, -5, 8, -3]))

[1, 3, 8]

Q5. 234라는 10진수의 16진수는 다음과 같이 구할 수 있다.  

    >>> hex(234)  
    '0xea'
이번에는 반대로 16진수 문자열 0xea를 10진수로 변경해 보자.

In [12]:
int('0xea', 16)

234

Q6. map과 lambda를 사용하여 [1, 2, 3, 4] 리스트의 각 요솟값에 3이 곱해진 리스트 [3, 6, 9, 12]를 만들어 보자.

In [13]:
list(map(lambda x:x*3, [1,2,3,4]))

[3, 6, 9, 12]

Q7. 다음 리스트의 최댓값과 최솟값의 합을 구해 보자.  
[-8, 2, 7, 5, -3, 5, 0, 1]

In [16]:
a = [-8, 2, 7, 5, -3, 5, 0, 1]
max(a) + min(a)

-1

Q8. 17/3의 결과는 다음과 같다.

In [17]:
17/3

5.666666666666667

위와 같은 결괏값 5.666666666666667을 소숫점 4자리까지만 반올림하여 표시해 보자.

In [18]:
round(17/3, 4)

5.6667

Q9. 다음과 같이 실행할 때 입력값을 모두 더하여 출력하는 스크립트를 작성해보자.

In [21]:
%%writefile myargv.py
import sys

numbers = sys.argv[1:]

result = 0
for number in numbers:
    result += int(number)
print(result)

Writing myargv.py


In [25]:
%ls

myargv.py  sample_data/


In [24]:
%run myargv.py 1 2 3 4 5 6 7 8 9 10

55


 Q10. os 모듈을 사용하여 다음과 같이 동작하도록 코드를 작성해 보자.  
 1. 해당 디렉토리로 이동한다.  
 2. dir 명령을 실행하고 그 결과를 변수에 담는다.
 3. dir 명령의 결과를 출력한다.

In [31]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [34]:
import os
os. chdir("/content/drive/My Drive/doit")
result = os.popen("dir")
print(result.read())

새파일2.txt  Hello.py  opener.py    readline.py  Shin.py
새파일3.txt  mod1.py   __pycache__  sample.py	 새파일.txt



Q11. glob 모듈을 이용하여 디렉토리의 파일 중 확장자가 .py인 파일만 출력하는 프로그램을 작성해보자.

In [35]:
import glob
glob.glob("/content/drive/My Drive/doit/*.py")


['/content/drive/My Drive/doit/readline.py',
 '/content/drive/My Drive/doit/mod1.py',
 '/content/drive/My Drive/doit/sample.py',
 '/content/drive/My Drive/doit/opener.py',
 '/content/drive/My Drive/doit/Shin.py',
 '/content/drive/My Drive/doit/Hello.py']

Q12. time 모듈을 사용하여 현재 날짜와 시간을 다음과 같은 형식으로 추력해보자.

In [36]:
import time
time.strftime("%Y/%m/%d %H:%M:%S")

'2019/10/09 09:13:24'

Q13. random 모듈을 사용하여 로또 번호(1~45 사이의 숫자 6개)를 생성해 보자. (단 중복된 숫자가 있으면 안 됨.)

In [37]:
import random
 
result = []
while len(result) < 6:
    num = random.randint(1, 45)
    if num not in result:
        result.append(num)

print(result)

[7, 5, 3, 15, 43, 36]


#### __init__.py의 용도  
* \_\_init__.py 파일은 해당 디렉토리가 패키지의 일부임을 알려주는 역할입니다.  
디렉터리에 \_\_init__.py 파일이 없다면 패키지로 인식되지 않을 수도 있습니다.  

#### __all__의 용도  
* 패키지의 \_\_init__.py에서 from .모듈 import 변수, 함수, 클래스 또는 from .모듈 import * 형식으로 작성했다면  
 패키지.함수() 형식으로 사용할 수 있습니다(변수, 클래스도 같은 형식).  
* 패키지의 \_\_init\_\_.py에서 from .모듈 import \*로 모든 변수, 함수, 클래스를 가져오면 패키지 외부에 공개하고 싶지 않은 것까지 공개하게 됩니다. 이때는 \_\_all__에 공개할 모듈, 변수, 함수, 클래스를 리스트 형태로 지정해주면 됩니다. \_\_all\_\_이라는 이름 그대로 모든 것(*)을 가져갈 때의 목록을 정합니다.

> C:/doit/game/sound/\_\_init\_\_.py  파일  
\_\_all\_\_=['내가 원하는 변수', '내가 원하는 함수', '내가 원하는 클래스']